src/crab/cow.hpp

```
@@ -0,0 +1,46 @@
+
+/// Copy-on-write wrapper. Const access (`*`, `->`) is free (shared);
+/// mutable access (`get_mutable()`) detaches when the underlying object
+/// is shared with other Cow instances.
+///
+/// Invariant: the internal shared_ptr is never null.
+template <typename T>
+class Cow final {
+    std::shared_ptr<T> ptr_;
+
+  public:
+    explicit Cow(std::shared_ptr<T> p) : ptr_(std::move(p)) { assert(ptr_); }
```

Clang (14.0.6)

[warning] constructor does not initialize these fields: ptr_

(cppcoreguidelines-pro-type-member-init)